ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -608,16 +614,54 @@ static void HandleKernelRequest(void* messageMemory, uint32_t messageSize)
         case MessageType_KtoU_NotifyDirectoryRenamed:
         case MessageType_KtoU_NotifyFileHardLinkCreated:
         {
+            // Walk up the directory tree and notify the provider about any directories
+            // not flagged as being in the root
+            stack<pair<string /*relative path*/, string /*full path*/>> newFolderPaths;
+            string parentPath(request.path);
+            size_t lastDirSeparator = parentPath.find_last_of('/');
+            while (lastDirSeparator != string::npos && lastDirSeparator > 0)
+            {
+                parentPath = parentPath.substr(0, lastDirSeparator);
+                char parentFullPath[PrjFSMaxPath];
+                CombinePaths(s_virtualizationRootFullPath.c_str(), parentPath.c_str(), parentFullPath);
+                if (IsBitSetInFileFlags(parentFullPath, FileFlags_IsInVirtualizationRoot))
+                {
+                    break;
+                }
+                else
+                {
+                    newFolderPaths.emplace(make_pair(parentPath, parentFullPath));
+                    lastDirSeparator = parentPath.find_last_of('/');
+                }
+            }
+            
+            while (!newFolderPaths.empty())
+            {
+                const pair<string /*relative path*/, string /*full path*/>& parentFolderPath = newFolderPaths.top();
+                newFolderPaths.pop();
```

Comment:
Don't pop until the bottom of the code block when we're done using `parentFolderPath`